### PR TITLE
[Move] Telekinesis + [Bug] Ingrain

### PR DESCRIPTION
**Pull request by frutescens**
## What are the changes the user will see?
Telekinesis will now be usable. 
Ingrain also properly grounds the user. 

## Why am I making these changes?
Gotta implement all the moves. 

## What are the changes from a developer perspective?
New Battler Tag for Telekinesis that governs the 100% hit chance.
Renaming MagnetRisenTag to FloatingTag so it is less specific and can be applied to the effects of Telekinesis. 
Gravity and Ingrain + Smack Down/1000 Arrows will have some related changes. 
Ingrain did not add the IGNORE_FLYING tag to the Pokemon affected - this PR changes that. 

Checklist:
- [x] Telekinesis raises its target into the air for three turns. The target will be immune to Ground-type moves and Arena Trap. 
- [x] This makes all moves used against the target (with the exception of one-hit knockout moves) bypass accuracy checks to always hit, unless the target is in the semi-invulnerable turn of a move such as Dig or Fly.
- [x]  If a Pokémon under the effect of Telekinesis is affected by Ingrain or holds an Iron Ball, it becomes grounded (i.e. it becomes vulnerable to Ground-type moves, etc.), but the effect to make moves targeting it ignore accuracy checks still applies. 
- [x] If a Pokémon under the effect of Telekinesis is hit by Smack Down or Thousand Arrows, the effect of Telekinesis disappears. 
- [x] If Gravity comes into effect, the effect of Telekinesis disappears from all Pokémon.
- [x] The move Telekinesis cannot affect a Pokémon under the effect of Ingrain, Smack Down, or Thousand Arrows. 
- [ ] While Gravity is in effect, Telekinesis cannot be selected.
- [x] Mega Gengar, Diglett, Dugtrio, Sandygast, and Palossand cannot be affected by the move Telekinesis, but a Pokémon transformed into one of these Pokémon can be affected. Telekinesis can also be baton passed to these Pokemon. 

PokeRogue Specific Changes
- [x] Telekinesis can now be Baton Passed to Mega Gengar
- [x] Wiglett/Wugtrio cannot be affected by the move Telekinesis.  

### Screenshots/Videos

https://github.com/user-attachments/assets/d16fcd08-07c1-4854-9f8b-b14297b2e4cc

## How to test the changes?
Use Telekinesis + Test the many edge cases involved.

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [x] If I have text, did I make it translatable and add a key in the English locale file(s)? https://github.com/pagefaultgames/pokerogue-locales/pull/16
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/data/move.ts [hunk 1]
```diff
@@ -8385,17 +8386,20 @@ export function initMoves() {
     new SelfStatusMove(Moves.RAGE_POWDER, Type.BUG, -1, 20, -1, 2, 5)
       .powderMove()
       .attr(AddBattlerTagAttr, BattlerTagType.CENTER_OF_ATTENTION, true),
-    new StatusMove(Moves.TELEKINESIS, Type.PSYCHIC, -1, 15, -1, 0, 5)
+    new StatusMove(Moves.TELEKINESIS, Type.PSYCHIC, 100, 15, -1, 0, 5)
       .condition(failOnGravityCondition)
-      .unimplemented(),
+      .condition((_user, target, _move) => ![ Species.DIGLETT, Species.DUGTRIO, Species.ALOLA_DIGLETT, Species.ALOLA_DUGTRIO, Species.SANDYGAST, Species.PALOSSAND, Species.WIGLETT, Species.WUGTRIO ].includes(target.species.speciesId))
+      .condition((_user, target, _move) => (target.species.speciesId !== Species.GENGAR && target.getFormKey() !== "mega"))
+      //.condition((_user, target, _move) => target.getTag(BattlerTagType.INGRAIN) || target.getTag(BattlerTagType.IGNORE_FLYING)
```
Comment (DayKev) [inline src/data/move.ts:8393]:
```suggestion
```

File: src/test/moves/telekinesis.test.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  it("Telekinesis makes the affected vulnerable to most attacking moves regardless of accuracy", async () => {
+    await game.classicMode.startBattle([ Species.MAGIKARP ]);
+
+    const enemyOpponent = game.scene.getEnemyPokemon();
+    expect(enemyOpponent).toBeDefined();
+
+    game.move.select(Moves.TELEKINESIS);
+    await game.phaseInterceptor.to("TurnEndPhase");
+    expect(enemyOpponent?.getTag(BattlerTagType.TELEKINESIS)).toBeDefined();
+    expect(enemyOpponent?.getTag(BattlerTagType.FLOATING)).toBeDefined();
```
Comment (DayKev) [inline src/test/moves/telekinesis.test.ts:45]:
```suggestion
    expect(enemyOpponent.getTag(BattlerTagType.TELEKINESIS)).toBeDefined();
    expect(enemyOpponent.getTag(BattlerTagType.FLOATING)).toBeDefined();
```

File: src/test/moves/telekinesis.test.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    game.move.select(Moves.TACKLE);
+    await game.phaseInterceptor.to("TurnEndPhase");
+    expect(enemyOpponent?.hp).toBeLessThan(enemyOpponent?.getMaxHp()!);
+  });
+
+  it("Telekinesis makes the affected airborne and immune to most Ground-moves", async () => {
+    await game.classicMode.startBattle([ Species.MAGIKARP ]);
+
+    const enemyOpponent = game.scene.getEnemyPokemon();
+    expect(enemyOpponent).toBeDefined();
```
Comment (DayKev) [inline src/test/moves/telekinesis.test.ts:58]:
```suggestion
    const enemyOpponent = game.scene.getEnemyPokemon()!;
```

File: src/test/moves/telekinesis.test.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+    game.move.select(Moves.TELEKINESIS);
+    await game.phaseInterceptor.to("TurnEndPhase");
+    expect(enemyOpponent?.getTag(BattlerTagType.TELEKINESIS)).toBeDefined();
+    expect(enemyOpponent?.getTag(BattlerTagType.FLOATING)).toBeDefined();
```
Comment (DayKev) [inline src/test/moves/telekinesis.test.ts:63]:
```suggestion
    expect(enemyOpponent.getTag(BattlerTagType.TELEKINESIS)).toBeDefined();
    expect(enemyOpponent.getTag(BattlerTagType.FLOATING)).toBeDefined();
```

File: src/test/moves/telekinesis.test.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+    await game.toNextTurn();
+    vi.spyOn(allMoves[Moves.MUD_SHOT], "accuracy", "get").mockReturnValue(100);
+    game.move.select(Moves.MUD_SHOT);
+    await game.phaseInterceptor.to("TurnEndPhase");
+    expect(enemyOpponent?.hp).toEqual(enemyOpponent?.getMaxHp()!);
```
Comment (DayKev) [inline src/test/moves/telekinesis.test.ts:69]:
```suggestion
    expect(enemyOpponent.isFullHp()).toBe(true);
```

File: src/test/moves/telekinesis.test.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+  });
+
+  it("Telekinesis can still affect Pokemon that have been transformed into invalid Pokemon", async () => {
+    game.override.enemyMoveset(Moves.TRANSFORM);
+    await game.classicMode.startBattle([ Species.DIGLETT ]);
+
+    const enemyOpponent = game.scene.getEnemyPokemon();
+    expect(enemyOpponent).toBeDefined();
```
Comment (DayKev) [inline src/test/moves/telekinesis.test.ts:77]:
```suggestion
    const enemyOpponent = game.scene.getEnemyPokemon()!;
```

File: src/test/moves/telekinesis.test.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+    game.move.select(Moves.TELEKINESIS);
+    await game.phaseInterceptor.to("TurnEndPhase");
+    expect(enemyOpponent?.getTag(BattlerTagType.TELEKINESIS)).toBeDefined();
+    expect(enemyOpponent?.getTag(BattlerTagType.FLOATING)).toBeDefined();
+    expect(enemyOpponent?.summonData.speciesForm?.speciesId).toBe(Species.DIGLETT);
```
Comment (DayKev) [inline src/test/moves/telekinesis.test.ts:83]:
```suggestion
    expect(enemyOpponent.getTag(BattlerTagType.TELEKINESIS)).toBeDefined();
    expect(enemyOpponent.getTag(BattlerTagType.FLOATING)).toBeDefined();
    expect(enemyOpponent.summonData.speciesForm?.speciesId).toBe(Species.DIGLETT);
```

File: src/test/moves/telekinesis.test.ts [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  });
+
+  it("Moves like Smack Down and 1000 Arrows remove all effects of Telekinesis from the target Pokemon", async () => {
+    await game.classicMode.startBattle([ Species.MAGIKARP ]);
+
+    const enemyOpponent = game.scene.getEnemyPokemon();
+    expect(enemyOpponent).toBeDefined();
```
Comment (DayKev) [inline src/test/moves/telekinesis.test.ts:90]:
```suggestion
    const enemyOpponent = game.scene.getEnemyPokemon()!;
```

File: src/test/moves/telekinesis.test.ts [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+    game.move.select(Moves.TELEKINESIS);
+    await game.phaseInterceptor.to("TurnEndPhase");
+    expect(enemyOpponent?.getTag(BattlerTagType.TELEKINESIS)).toBeDefined();
+    expect(enemyOpponent?.getTag(BattlerTagType.FLOATING)).toBeDefined();
```
Comment (DayKev) [inline src/test/moves/telekinesis.test.ts:95]:
```suggestion
    expect(enemyOpponent.getTag(BattlerTagType.TELEKINESIS)).toBeDefined();
    expect(enemyOpponent.getTag(BattlerTagType.FLOATING)).toBeDefined();
```

File: src/test/moves/telekinesis.test.ts [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+    await game.toNextTurn();
+    game.move.select(Moves.SMACK_DOWN);
+    await game.phaseInterceptor.to("TurnEndPhase");
+    expect(enemyOpponent?.getTag(BattlerTagType.TELEKINESIS)).toBeUndefined();
+    expect(enemyOpponent?.getTag(BattlerTagType.FLOATING)).toBeUndefined();
```
Comment (DayKev) [inline src/test/moves/telekinesis.test.ts:101]:
```suggestion
    expect(enemyOpponent.getTag(BattlerTagType.TELEKINESIS)).toBeUndefined();
    expect(enemyOpponent.getTag(BattlerTagType.FLOATING)).toBeUndefined();
```

File: src/test/moves/telekinesis.test.ts [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+  });
+
+  it("Ingrain will remove the floating effect of Telekinesis, but not the 100% hit", async () => {
+    game.override.enemyMoveset([ Moves.SPLASH, Moves.INGRAIN ]);
+    await game.classicMode.startBattle([ Species.MAGIKARP ]);
+
+    const playerPokemon = game.scene.getPlayerPokemon();
+    const enemyOpponent = game.scene.getEnemyPokemon();
+    expect(enemyOpponent).toBeDefined();
```
Comment (DayKev) [inline src/test/moves/telekinesis.test.ts:110]:
```suggestion
    const playerPokemon = game.scene.getPlayerPokemon()!;
    const enemyOpponent = game.scene.getEnemyPokemon()!;
```

File: src/test/moves/telekinesis.test.ts [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+    game.move.select(Moves.TELEKINESIS);
+    await game.forceEnemyMove(Moves.SPLASH);
+    await game.phaseInterceptor.to("TurnEndPhase");
+    expect(enemyOpponent?.getTag(BattlerTagType.TELEKINESIS)).toBeDefined();
+    expect(enemyOpponent?.getTag(BattlerTagType.FLOATING)).toBeDefined();
```
Comment (DayKev) [inline src/test/moves/telekinesis.test.ts:116]:
```suggestion
    expect(enemyOpponent.getTag(BattlerTagType.TELEKINESIS)).toBeDefined();
    expect(enemyOpponent.getTag(BattlerTagType.FLOATING)).toBeDefined();
```

File: src/test/moves/telekinesis.test.ts [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    vi.spyOn(allMoves[Moves.MUD_SHOT], "accuracy", "get").mockReturnValue(0);
+    game.move.select(Moves.MUD_SHOT);
+    await game.forceEnemyMove(Moves.INGRAIN);
+    await game.phaseInterceptor.to("TurnEndPhase");
+    expect(enemyOpponent?.getTag(BattlerTagType.TELEKINESIS)).toBeDefined();
+    expect(enemyOpponent?.getTag(BattlerTagType.INGRAIN)).toBeDefined();
+    expect(enemyOpponent?.getTag(BattlerTagType.IGNORE_FLYING)).toBeDefined();
+    expect(enemyOpponent?.getTag(BattlerTagType.FLOATING)).toBeUndefined();
+    const lastMoveResult = playerPokemon?.getLastXMoves()[0].result;
+    expect(lastMoveResult).toBe(MoveResult.SUCCESS);
```
Comment (DayKev) [inline src/test/moves/telekinesis.test.ts:128]:
```suggestion
    expect(enemyOpponent.getTag(BattlerTagType.TELEKINESIS)).toBeDefined();
    expect(enemyOpponent.getTag(BattlerTagType.INGRAIN)).toBeDefined();
    expect(enemyOpponent.getTag(BattlerTagType.IGNORE_FLYING)).toBeDefined();
    expect(enemyOpponent.getTag(BattlerTagType.FLOATING)).toBeUndefined();
    expect(playerPokemon.getLastXMoves()[0].result).toBe(MoveResult.SUCCESS);
```

File: src/data/move.ts (hunk 1 above)
Comment (frutescens) [reply src/data/move.ts]: wait that was not supposed to be like that

**Review by DayKev — APPROVED** [review]

File: src/data/move.ts [hunk 14]
```diff
@@ -8385,17 +8386,20 @@ export function initMoves() {
     new SelfStatusMove(Moves.RAGE_POWDER, Type.BUG, -1, 20, -1, 2, 5)
       .powderMove()
       .attr(AddBattlerTagAttr, BattlerTagType.CENTER_OF_ATTENTION, true),
-    new StatusMove(Moves.TELEKINESIS, Type.PSYCHIC, -1, 15, -1, 0, 5)
+    new StatusMove(Moves.TELEKINESIS, Type.PSYCHIC, 100, 15, -1, 0, 5)
```
Comment (PigeonBar) [inline src/data/move.ts:8389]:
```suggestion
    new StatusMove(Moves.TELEKINESIS, Type.PSYCHIC, -1, 15, -1, 0, 5)
```
According to [Bulbapedia](https://bulbapedia.bulbagarden.net/wiki/Telekinesis_(move)), Telekinesis should bypass accuracy checks.

File: src/data/battler-tags.ts [hunk 15]
```diff
@@ -2671,6 +2685,24 @@ export class SyrupBombTag extends BattlerTag {
...
 }
 
+export class TelekinesisTag extends BattlerTag {
+  constructor(sourceMove: Moves) {
+    super(BattlerTagType.TELEKINESIS, [ BattlerTagLapseType.PRE_MOVE, BattlerTagLapseType.AFTER_MOVE ], 3, sourceMove, undefined, true);
+  }
+
+  override onAdd(pokemon: Pokemon)  {
+    pokemon.scene.queueMessage(i18next.t("battlerTags:telekinesisOnAdd", { pokemonNameWithAffix: getPokemonNameWithAffix(pokemon) }));
+    pokemon.addTag(BattlerTagType.TELEKINESIS, 3, this.sourceMove);
```
Comment (innerthunder) [inline src/data/battler-tags.ts:2695]: `addTag` calls `onAdd`, so this line isn't needed

File: src/data/battler-tags.ts [hunk 16]
```diff
@@ -2671,6 +2685,24 @@ export class SyrupBombTag extends BattlerTag {
...
+    pokemon.addTag(BattlerTagType.FLOATING, 3, this.sourceMove);
+  }
+
+  override onRemove(pokemon: Pokemon) {
+    if (pokemon.getTag(BattlerTagType.FLOATING)) {
+      pokemon.removeTag(BattlerTagType.FLOATING);
+    }
```
Comment (innerthunder) [inline src/data/battler-tags.ts:2702]:
```suggestion
    pokemon.removeTag(BattlerTagType.FLOATING);
```
Slight optimization; `removeTag` just doesn't do anything if a matching tag isn't found.

File: src/data/battler-tags.ts [hunk 17]
```diff
@@ -964,6 +964,11 @@ export class IngrainTag extends TrappedTag {
     super(BattlerTagType.INGRAIN, BattlerTagLapseType.TURN_END, 1, Moves.INGRAIN, sourceId);
   }
 
+  override onAdd(pokemon: Pokemon) {
+    super.onAdd(pokemon);
+    pokemon.addTag(BattlerTagType.IGNORE_FLYING, undefined, Moves.INGRAIN, this.sourceId);
+  }
```
Comment (innerthunder) [inline src/data/battler-tags.ts:970]: I think it'd be simpler if Ingrain just had 2 different `AddBattlerTagAttrs` to add these tags separately. It'd basically remove the need for this override and the `onRemove` override. You could probably get away with doing the same thing for Telekinesis, but I'm less sure on that.

File: src/data/move.ts [hunk 18]
```diff
@@ -8387,15 +8388,18 @@ export function initMoves() {
       .attr(AddBattlerTagAttr, BattlerTagType.CENTER_OF_ATTENTION, true),
     new StatusMove(Moves.TELEKINESIS, Type.PSYCHIC, -1, 15, -1, 0, 5)
       .condition(failOnGravityCondition)
-      .unimplemented(),
+      .condition((_user, target, _move) => ![ Species.DIGLETT, Species.DUGTRIO, Species.ALOLA_DIGLETT, Species.ALOLA_DUGTRIO, Species.SANDYGAST, Species.PALOSSAND, Species.WIGLETT, Species.WUGTRIO ].includes(target.species.speciesId))
+      .condition((_user, target, _move) => (target.species.speciesId !== Species.GENGAR && target.getFormKey() !== "mega"))
```
Comment (innerthunder) [inline src/data/move.ts:8392]:
```suggestion
      .condition((_user, target, _move) => !(target.species.speciesId === Species.GENGAR && target.getFormKey() === "mega"))
```
The old condition causes TK to fail against all Mega-Evos

File: src/field/pokemon.ts [hunk 19]
```diff
@@ -1488,7 +1488,7 @@ export default abstract class Pokemon extends Phaser.GameObjects.Container {
   }
 
   isGrounded(): boolean {
-    return !!this.getTag(GroundedTag) || (!this.isOfType(Type.FLYING, true, true) && !this.hasAbility(Abilities.LEVITATE) && !this.getTag(BattlerTagType.MAGNET_RISEN) && !this.getTag(SemiInvulnerableTag));
+    return !!this.getTag(GroundedTag) || (!this.isOfType(Type.FLYING, true, true) && !this.hasAbility(Abilities.LEVITATE) && !this.getTag(BattlerTagType.FLOATING) && !this.getTag(SemiInvulnerableTag));
```
Comment (innerthunder) [inline src/field/pokemon.ts:1491]:
```suggestion
    return !!this.getTag(GroundedTag) || (!this.isOfType(Type.FLYING, true, true) && !this.hasAbility(Abilities.LEVITATE) && !this.getTag(BattlerTagType.FLOATING) && !this.getTag(BattlerTagType.FLYING));
```

I feel like using the entire `SemiInvulnerableTag` class here introduces issues...

**Review by innerthunder — CHANGES_REQUESTED** [review]

File: src/field/pokemon.ts (hunk 19 above)
Comment (frutescens) [reply src/field/pokemon.ts:1491]: I didn't add the check for SemiInvulnerableTag in this PR so this change is out of scope for this PR.

File: src/data/battler-tags.ts (hunk 17 above)
Comment (frutescens) [reply src/data/battler-tags.ts:970]: I disagree since this shows that one tag is dependent on the other tag's presence. Same with Telekinesis since there are situations where a Pokemon can be floating, but still subject to Telekinesis.

Comment (frutescens) [reply src/data/battler-tags.ts]: Yeah, I think it should be fine to move them as part of move attr.

File: src/data/battler-tags.ts [hunk 20]
```diff
@@ -964,6 +964,10 @@ export class IngrainTag extends TrappedTag {
     super(BattlerTagType.INGRAIN, BattlerTagLapseType.TURN_END, 1, Moves.INGRAIN, sourceId);
   }
 
+  override onAdd(pokemon: Pokemon) {
+    super.onAdd(pokemon);
+  }
+
```
Comment (innerthunder) [inline src/data/battler-tags.ts:970]:
```suggestion
```

File: src/data/battler-tags.ts [hunk 21]
```diff
@@ -993,6 +997,11 @@ export class IngrainTag extends TrappedTag {
     return ret;
   }
 
+  override onRemove(pokemon: Pokemon) {
+    super.onRemove(pokemon);
+    pokemon.removeTag(BattlerTagType.IGNORE_FLYING);
+  }
+
```
Comment (innerthunder) [inline src/data/battler-tags.ts:1004]: I don't think this is really needed, either (?). `resetSummonData`'s basically the only thing that removes these tags, anyway (well, that and fainting), and both events remove all tags in general

Comment (frutescens) [reply src/data/battler-tags.ts]: Fixed.

**Review by innerthunder — APPROVED** [review]

File: src/data/battler-tags.ts [hunk 22]
```diff
@@ -2676,6 +2680,16 @@ export class SyrupBombTag extends BattlerTag {
   }
 }
 
+export class TelekinesisTag extends BattlerTag {
```
Comment (Tempo-anon) [inline src/data/battler-tags.ts:2683]: Since you're here, mind documenting FloatingTag and TelekinesisTag?

File: src/data/move.ts [hunk 23]
```diff
@@ -8402,15 +8404,19 @@ export function initMoves() {
       .attr(AddBattlerTagAttr, BattlerTagType.CENTER_OF_ATTENTION, true),
     new StatusMove(Moves.TELEKINESIS, Type.PSYCHIC, -1, 15, -1, 0, 5)
       .condition(failOnGravityCondition)
-      .unimplemented(),
+      .condition((_user, target, _move) => ![ Species.DIGLETT, Species.DUGTRIO, Species.ALOLA_DIGLETT, Species.ALOLA_DUGTRIO, Species.SANDYGAST, Species.PALOSSAND, Species.WIGLETT, Species.WUGTRIO ].includes(target.species.speciesId))
+      .condition((_user, target, _move) => !(target.species.speciesId === Species.GENGAR && target.getFormKey() === "mega"))
```
Comment (Tempo-anon) [inline src/data/move.ts:8408]: Might want to add a comment about the known edge case of Gengar becoming mega but all this hardcoded inability to be affected thing is honestly quite silly and I would not be opposed to just removing it altogether. Not like anyone ever uses this move any way.

Comment (frutescens) [reply src/data/move.ts:8408]: Ok, I'll add a comment there.

File: src/data/battler-tags.ts [hunk 24]
```diff
@@ -1713,21 +1713,31 @@ export class TypeImmuneTag extends BattlerTag {
   }
 }
 
-export class MagnetRisenTag extends TypeImmuneTag {
+/**
+ * Battler Tag that lifts the affected in the air and provides immunity to Ground Type Moves
+ * Moves that provide this effect are Magnet Rise and Telekinesis
```
Comment (DayKev) [inline src/data/battler-tags.ts:1718]:
```suggestion
 * Battler Tag that lifts the affected Pokemon into the air and provides immunity to Ground type moves.
```

**Review by DayKev — APPROVED** [review]

**Review by innerthunder — APPROVED** [review]
